identity: Fix nil pointer panic in LookupIdentityByID

gandro opened the pull request:
Because the identity allocator is initialized asynchronously via
`InitIdentityAllocator`, its internal local identitiy allocator might
not have been initialized yet when the lookup functions are called.
This can cause nil pointer panics, as observed in #13479.

Before b194612c004c3e69289286e9a35d337b2645fc50, this nil pointer panic could not occur in
`LookupIdentityByID` as the function checked for `m.IdentityAllocator != nil`
which also implies `m.localIdentities != nil`.
This commit adds an explict check for `m.localIdentities`.

Fixes: #13479
Fixes: b194612c004c ("identity: Avoid kvstore lookup for local identities")

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

---

Note: While the backported version of b194612c004c in v1.7 and v1.8 does not suffer from the nil pointer panic due to an early check fo `m.IdentityAllocator != nil` , I think we should still back-port this to fix the potential nil-pointer panic in `LookupIdentity` which is also fixed by this commit.

gandro committed:
identity: Fix nil pointer panic in LookupIdentityByID

Because the identity allocator is initialized asychronously via
`InitIdentityAllocator`, the local identitiy allocator might not have
been initialized yet when the lookup functions are called. This can
cause nil pointer panics, as observed in #13479.

Before b194612c004c3e69289286e9a35d337b2645fc50, this nil pointer panic
could not occur in `LookupIdentityByID` as the function checked for
`m.IdentityAllocator != nil` which also implies `m.localIdentities != nil`.

This commit adds an explict check for `m.localIdentities` and fixes a
potential data race by checking the initialization channels before
accessing `m.localIdentities` or `m.IdentityAllocator`.

Fixes: #13479
Fixes: b194612c004c ("identity: Avoid kvstore lookup for local identities")

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

aanm reviewed (changes requested):
@gandro why don't we need to wait before the variable is assigned with the `<-m.localIdentityAllocatorInitialized`?

gandro committed:
identity: Fix potential data races and nil pointer panics

The local and global allocator are initialized asynchronously in
`InitIdentityAllocator`. Therefore we must check for the initialization
channels a potential synchronization point before we can check for
`m.{localIdentities,IdentityAllocator} != nil`.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
